community/kernel/src/test/java/org/neo4j/kernel/impl/transaction/log/TransactionLogAppendAndRotateStressIT.java

```diff
@@ -0,0 +1,276 @@
+
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertTrue;
+import static org.junit.Assert.fail;
+import static org.junit.rules.RuleChain.outerRule;
+import static org.mockito.Mockito.mock;
+
+import static java.lang.System.currentTimeMillis;
+import static java.util.concurrent.TimeUnit.MILLISECONDS;
+import static java.util.concurrent.TimeUnit.SECONDS;
+import static java.util.concurrent.locks.LockSupport.parkNanos;
+
+import static org.neo4j.io.ByteUnit.mebiBytes;
+import static org.neo4j.kernel.impl.transaction.log.entry.LogHeader.LOG_HEADER_SIZE;
+import static org.neo4j.kernel.impl.transaction.tracing.LogAppendEvent.NULL;
+import static org.neo4j.kernel.impl.util.IdOrderingQueue.BYPASS;
+
+public class TransactionLogAppendAndRotateStressIT
+{
+    private final LifeRule life = new LifeRule( true );
+    private final TestDirectory directory = TargetDirectory.testDirForTest( getClass() );
+    @Rule
+    public final RuleChain chain = outerRule( directory ).around( life );
+
+    @Test
+    public void shouldKeepTransactionsIntactWhenConcurrentlyRotationAndAppending() throws Throwable
+    {
+        // GIVEN
+        FileSystemAbstraction fs = new DefaultFileSystemAbstraction();
+        PhysicalLogFiles logFiles = new PhysicalLogFiles( directory.directory().getAbsoluteFile(), fs );
+        long rotationThreshold = mebiBytes( 1 );
+        LogVersionRepository logVersionRepository = new DeadSimpleLogVersionRepository( 0 );
+        final AtomicBoolean end = new AtomicBoolean();
+        AllTheMonitoring monitoring = new AllTheMonitoring( end, 100 );
+        TransactionIdStore txIdStore = new DeadSimpleTransactionIdStore();
+        TransactionMetadataCache metadataCache = new TransactionMetadataCache( 10, 100 );
+        LogFile logFile = life.add( new PhysicalLogFile( fs, logFiles, rotationThreshold,
+                txIdStore, logVersionRepository, monitoring, metadataCache ) );
+        monitoring.setLogFile( logFile );
+        KernelHealth health = new KernelHealth( mock( KernelPanicEventGenerator.class ), NullLog.getInstance() );
+        LogRotation rotation = new LogRotationImpl( monitoring, logFile, health );
+        final TransactionAppender appender = life.add( new BatchingTransactionAppender( logFile, rotation, metadataCache,
+                txIdStore, BYPASS, health ) );
+
+        // WHEN
+        Race race = new Race();
+        for ( int i = 0; i < 10; i++ )
+        {
+            race.addContestant( new Runnable()
+            {
+                @Override
+                public void run()
+                {
+                    while ( !end.get() )
+                    {
+                        try
+                        {
+                            appender.append( sillyTransaction( 1_000 ), NULL );
+                        }
+                        catch ( Exception e )
+                        {
+                            e.printStackTrace( System.out );
```

Comment:
I think we can remove this printStackTrace since a test failure will show the exception anyway.